core/drivers/crypto/caam/caam_jr.c

```diff
@@ -0,0 +1,768 @@
+
+	struct inring_entry *cur_inrings = NULL;
+	struct caller_info *caller = NULL;
+	uint32_t exceptions = 0;
+	uint32_t job_mask = 0;
+	uint8_t  idx_jr = 0;
+	bool     found = false;
+
+	exceptions = cpu_spin_lock_xsave(&jr_privdata->inlock);
+
+	/* Stay lock until a job is available */
+	/* Check if there is an available JR index in the HW */
+	while (hal_jr_read_nbSlotAvailable(jr_privdata->baseaddr) == 0) {
+		/*
+		 * WFE will return thanks to a SEV generated by the
+		 * interrupt handler or by a spin_unlock
+		 */
+		wfe();
+	};
+
+	/*
+	 * There is a space free in the input ring but it doesn't mean
+	 * that the job pushed is completed.
+	 * Completion is out of order. Look for a free space in the
+	 * caller data to push them and get a job id for the completion
+	 *
+	 * Lock the caller information array because dequeue is
+	 * also touching it
+	 */
+	cpu_spin_lock(&jr_privdata->callers_lock);
+	for ((idx_jr = 0), (found = false);
+		(idx_jr < jr_privdata->nbJobs) && (!found); idx_jr++) {
+		if (jr_privdata->callers[idx_jr].jobid == JR_JOB_FREE) {
+			JR_TRACE("Found a space #%d free in the callers array",
+				idx_jr);
+			job_mask = 1 << idx_jr;
+
+			/* Store the caller information for the JR completion */
+			caller = &jr_privdata->callers[idx_jr];
+			caller->jobid  = job_mask;
+			caller->jobctx = jobctx;
+			caller->pdesc  = virt_to_phys((void *)jobctx->desc);
+
+			found = true;
+		}
+	}
+	cpu_spin_unlock(&jr_privdata->callers_lock);
+
+	if (!found) {
+		JR_TRACE("Error didn't find a free space in the callers array");
+		goto end_enqueue;
+	}
+
+	JR_TRACE("Push id=%d, job (0x%08x) context @0x%08"PRIxVA"",
+		jr_privdata->inwrite_index, job_mask, (vaddr_t)jobctx);
+
+	cur_inrings = &jr_privdata->inrings[jr_privdata->inwrite_index];
+
+	/* Push the descriptor into the JR HW list */
+	caam_write_val(cur_inrings, caller->pdesc);
+
+	/* Ensure that physical memory is up to date */
+	cache_operation(TEE_CACHECLEAN, (void *)cur_inrings,
+			sizeof(struct inring_entry));
+
+	/*
+	 * Increment index to next JR input entry taking care that
+	 * is a circular buffer of nbJobs size.
+	 */
+	jr_privdata->inwrite_index++;
+	jr_privdata->inwrite_index %= jr_privdata->nbJobs;
+
+	/* Ensure that input descriptor is pushed in physical memory */
+	cache_operation(TEE_CACHECLEAN, (void *)jobctx->desc,
+			DESC_SZBYTES(desc_get_len(jobctx->desc)));
+
+	/* Inform HW that a new JR is available */
+	hal_jr_add_newjob(jr_privdata->baseaddr);
+
+	*jobId = job_mask;
+	retstatus = CAAM_NO_ERROR;
+
+end_enqueue:
+	cpu_spin_unlock_xrestore(&jr_privdata->inlock, exceptions);
+
+	return retstatus;
+}
+
+/*
+ * Synchronous job completion callback
+ *
+ * @jobctx   Job context
+ */
+static void job_done(struct jr_jobctx *jobctx)
+{
+	jobctx->completion = true;
+}
+
+/*
+ * Cancels a job id. Remove the job from SW Job array
+ *
+ * @jobId     Job Id
+ */
+void caam_jr_cancel(uint32_t jobId)
+{
+	uint8_t idx = 0;
+
+	JR_TRACE("Job cancel 0x%"PRIx32"", jobId);
```

Comment:
@etienne-lms, when I saw this with the lack of spaces this first time I thought the same thing. However, after seeing this for a while it is actually a bit easier to read. I guess it's because it's slightly more compact and also looks a bit more like a format directive.

I don't mind if we keep skipping the spaces around `PRI*` macros in strings, I've started to use this style myself.